src/StreamJsonRpc/MessageFormatterHelper.cs

```diff
@@ -0,0 +1,126 @@
+﻿
+namespace StreamJsonRpc
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using Microsoft;
+    using Microsoft.VisualStudio.Threading;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Class containing useful method to help on the implementation of message formatters.
+    /// </summary>
+    public class MessageFormatterHelper
+    {
+        /// <summary>
+        /// Special method name for progress notification.
+        /// </summary>
+        protected const string ProgressRequestSpecialMethod = "$/progress";
+
+        /// <summary>
+        /// Object used to lock the acces to <see cref="requestProgressMap"/> and <see cref="progressMap"/>.
+        /// </summary>
+        protected readonly object progressLock = new object();
+
+        /// <summary>
+        /// Dictionary used to map the request id to their progress id token so that the progress objects are cleaned after getting the final response.
+        /// </summary>
+        protected readonly Dictionary<long, long> requestProgressMap = new Dictionary<long, long>();
```

Comment:
>long [](start = 38, length = 4)

Can you include in the xml doc comments whether this collection only includes *outbound* requests? I see you're assuming here that `long` is the type for the message ID, which isn't true generally, but seems like a reasonable and necessary assumption for outbound requests.